### PR TITLE
fix(sdk): update git repo handling for failure cases and rename to gitlib

### LINKED ISSUE #5291: Wandb init fails when on inferring environment from Git
I am trying to run jobs via a Kubernetes cluster and my job fails whenever it reaches `wandb.init` with the following error:

```
wandb: ERROR Abnormal program exit
Traceback (most recent call last):
  File "/yes/envs/titan/lib/python3.9/site-packages/wandb/sdk/wandb_init.py", line 996, in init
    wi.setup(kwargs)
  File "/yes/envs/titan/lib/python3.9/site-packages/wandb/sdk/wandb_init.py", line 133, in setup
    self._wl = wandb_setup.setup()
  File "/yes/envs/titan/lib/python3.9/site-packages/wandb/sdk/wandb_setup.py", line 318, in setup
    ret = _setup(settings=settings)
  File "/yes/envs/titan/lib/python3.9/site-packages/wandb/sdk/wandb_setup.py", line 313, in _setup
    wl = _WandbSetup(settings=settings)
  File "/yes/envs/titan/lib/python3.9/site-packages/wandb/sdk/wandb_setup.py", line 299, in __init__
    _WandbSetup._instance = _WandbSetup__WandbSetup(settings=settings, pid=pid)
  File "/yes/envs/titan/lib/python3.9/site-packages/wandb/sdk/wandb_setup.py", line 107, in __init__
    self._settings = self._settings_setup(settings, self._early_logger)
  File "/yes/envs/titan/lib/python3.9/site-packages/wandb/sdk/wandb_setup.py", line 136, in _settings_setup
    s._infer_run_settings_from_environment(_logger=early_logger)
  File "/yes/envs/titan/lib/python3.9/site-packages/wandb/sdk/wandb_settings.py", line 1392, in _infer_run_settings_from_environment
    program_relpath = self.program_relpath or _get_program_relpath_from_gitrepo(
  File "/yes/envs/titan/lib/python3.9/site-packages/wandb/sdk/wandb_settings.py", line 138, in _get_program_relpath_from_gitrepo
    root = repo.root
  File "/yes/envs/titan/lib/python3.9/site-packages/wandb/sdk/lib/git.py", line 46, in root
    return self.repo.git.rev_parse("--show-toplevel")
  File "/yes/envs/titan/lib/python3.9/site-packages/git/cmd.py", line 639, in <lambda>
    return lambda *args, **kwargs: self._call_process(name, *args, **kwargs)
  File "/yes/envs/titan/lib/python3.9/site-packages/git/cmd.py", line 1184, in _call_process
    return self.execute(call, **exec_kwargs)
  File "/yes/envs/titan/lib/python3.9/site-packages/git/cmd.py", line 984, in execute
    raise GitCommandError(redacted_command, status, stderr_value, stdout_value)
git.exc.GitCommandError: Cmd('git') failed due to: exit code(128)
  cmdline: git rev-parse --show-toplevel
  stderr: 'fatal: detected dubious ownership in repository at '/nfs'
To add an exception for this directory, call:

	git config --global --add safe.directory /nfs'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/nfs/repr_comparison/scripts/source_training/train.py", line 136, in <module>
    run(args)
  File "/nfs/repr_comparison/scripts/source_training/train.py", line 84, in run
    wandb.init(
  File "/yes/envs/titan/lib/python3.9/site-packages/wandb/sdk/wandb_init.py", line 1037, in init
    raise Exception("problem") from error_seen
Exception: problem
```

/nfs/ here is a cluster-level name pointing to my specific Kubernetes volume claim.

I have tried passing in `wandb.Settings(program_relpath=os.path.join(os.getenv("SCRIPT_DIR"), "train.py"), disable_git=True, save_code=False)` (as per https://github.com/wandb/wandb/issues/1565) but it still seems to check for git when I run it.

Since I launch these in Docker containers, it's not as simple as me just running the suggested line of  `git config --global --add safe.directory /nfs`

### PR DESCRIPTION
**Pull request by kptkin**
Fixes
-----

Fixes WB-13236

Fixes #5291

Description
-----------
What does the PR do?

Some users have a `.git` repo on their system, that doesn't actually relevant for them, or other form of invalid git. Our current implementation tried to use this malformed git info to extract information relevant to this, and caused user's script to fail. This PR just hardens our git utility to catch the relevant exception thrown by the git library and return a default value in that case...

Testing
-------
How was this PR tested?

Tested locally with a malformed .git folder, it fails in earlier version and works in this latest fix.

Checklist
-------
- [x] Include reference to internal ticket "Fixes WB-NNNN" and/or GitHub issue "Fixes #NNNN" (if applicable)
- [x] Ensure PR title compliance with the [conventional commits standards](https://github.com/wandb/wandb/blob/main/CONTRIBUTING.md#conventional-commits)

### <samp>🤖 Generated by Copilot at 2829f36</samp>

> _`GitRepo` refined_
> _Using specific exceptions_
> _Debug code removed_

### CODE REVIEW

File: wandb/sdk/lib/git.py [hunk 1]
```diff
@@ -114,6 +122,7 @@ def remote(self):
         if not self.repo:
             return None
         try:
+            breakpoint()
```
Comment (kptkin) [inline wandb/sdk/lib/git.py:125]:
```suggestion
```

File: wandb/sdk/lib/git.py [hunk 2]
```diff
@@ -4,6 +4,9 @@
 from typing import Optional
 from urllib.parse import urlparse, urlunparse
 
+from git.exc import GitCommandError, InvalidGitRepositoryError, NoSuchPathError
```
Comment (kptkin) [inline wandb/sdk/lib/git.py:7]: undo changes

**Review by dmitryduev — DISMISSED** [review]

File: wandb/sdk/lib/git.py [hunk 3]
```diff
@@ -64,7 +68,8 @@ def root(self) -> Optional[str]:
             return None
         try:
             return self.repo.git.rev_parse("--show-toplevel")
-        except exc.GitCommandError as e:
+        except GitCommandError as e:
+            # raise Exception("git root error: {}".format(e))
```
Comment (kptkin) [inline wandb/sdk/lib/git.py:72]:
```suggestion
```

File: wandb/sdk/lib/gitilb.py [hunk 4]
```diff
@@ -90,6 +104,7 @@ def last_commit(self):
             return self._commit
         if not self.repo:
             return None
+        breakpoint()
```
Comment (kptkin) [inline wandb/sdk/lib/gitilb.py:107]:
```suggestion
```

File: tests/pytest_tests/conftest.py [hunk 5]
```diff
@@ -17,7 +17,7 @@
 from wandb import Api  # noqa: E402
 from wandb.sdk.interface.interface_queue import InterfaceQueue  # noqa: E402
 from wandb.sdk.lib import filesystem, runid  # noqa: E402
-from wandb.sdk.lib.git import GitRepo  # noqa: E402
+from wandb.sdk.lib.gitilb import GitRepo  # noqa: E402
```
Comment (antonblr) [inline tests/pytest_tests/conftest.py:20]:
```suggestion
from wandb.sdk.lib.gitlib import GitRepo  # noqa: E402
```
While it is inline with the imported file/module name, I think it is typo (here and elsewhere)?

Comment (kptkin) [reply tests/pytest_tests/conftest.py:20]: oh thanks this is actually correct

File: tests/pytest_tests/unit_tests/test_lib/test_git.py [hunk 6]
```diff
@@ -1,17 +1,15 @@
...
 """Tests for the `wandb.GitRepo` module."""
-from typing import Callable, Optional
+from typing import Generator, Optional
 
 import git
 import pytest
-import wandb
-
-GitRepo = wandb.wandb_lib.git.GitRepo
+from wandb.sdk.lib.gitilb import GitRepo
```
Comment (kptkin) [inline tests/pytest_tests/unit_tests/test_lib/test_git.py:8]:
```suggestion
from wandb.sdk.lib.gitlib import GitRepo
```

File: tests/pytest_tests/unit_tests_old/conftest.py [hunk 7]
```diff
@@ -35,7 +35,7 @@
 from wandb.sdk.internal.sender import SendManager
 from wandb.sdk.internal.writer import WriteManager
 from wandb.sdk.lib import filesystem, runid
-from wandb.sdk.lib.git import GitRepo
+from wandb.sdk.lib.gitilb import GitRepo
```
Comment (kptkin) [inline tests/pytest_tests/unit_tests_old/conftest.py:38]:
```suggestion
from wandb.sdk.lib.gitlib import GitRepo
```

File: wandb/sdk/internal/internal_api.py [hunk 8]
```diff
@@ -44,7 +44,7 @@
 
 from ..lib import retry
 from ..lib.filenames import DIFF_FNAME, METADATA_FNAME
-from ..lib.git import GitRepo
+from ..lib.gitilb import GitRepo
```
Comment (kptkin) [inline wandb/sdk/internal/internal_api.py:47]:
```suggestion
from ..lib.gitlib import GitRepo
```

File: wandb/sdk/wandb_settings.py [hunk 9]
```diff
@@ -43,7 +43,7 @@
 from wandb.sdk.wandb_setup import _EarlyLogger
 
 from .lib import apikey
-from .lib.git import GitRepo
+from .lib.gitilb import GitRepo
```
Comment (kptkin) [inline wandb/sdk/wandb_settings.py:46]:
```suggestion
from .lib.gitlib import GitRepo
```

File: wandb/util.py [hunk 10]
```diff
@@ -1267,7 +1267,9 @@ def from_human_size(size: str, units: Optional[List[Tuple[str, Any]]] = None) ->
 
 def auto_project_name(program: Optional[str]) -> str:
     # if we're in git, set project name to git repo name + relative path within repo
-    root_dir = wandb.wandb_sdk.lib.git.GitRepo().root_dir
+    from wandb.sdk.lib.gitilb import GitRepo
```
Comment (kptkin) [inline wandb/util.py:1270]:
```suggestion
    from wandb.sdk.lib.gitlib import GitRepo
```

File: wandb/sdk/wandb_run.py [hunk 11]
```diff
@@ -83,7 +83,7 @@
     telemetry,
 )
 from .lib.exit_hooks import ExitHooks
-from .lib.git import GitRepo
+from .lib.gitilb import GitRepo
```
Comment (kptkin) [inline wandb/sdk/wandb_run.py:86]:
```suggestion
from .lib.gitlib import GitRepo
```

**Review by dmitryduev — APPROVED** [review]

File: wandb/sdk/lib/gitlib.py [hunk 12]
```diff
@@ -104,16 +127,17 @@ def last_commit(self):
...
             return None
         return self.repo.head.ref.name
 
     @property
-    def remote(self):
+    def remote(self) -> Any:
         if not self.repo:
             return None
         try:
+            assert self.remote_name is not None
```
Comment (kptkin) [inline wandb/sdk/lib/gitlib.py:140]:
```suggestion
```

File: wandb/sdk/lib/gitlib.py [hunk 13]
```diff
@@ -104,13 +127,13 @@ def last_commit(self):
             return None
 
     @property
-    def branch(self) -> Optional[str]:
+    def branch(self) -> Any:
         if not self.repo:
             return None
         return self.repo.head.ref.name
```
Comment (kptkin) [inline wandb/sdk/lib/gitlib.py:133]: should we catch the assertion error as well?

Comment (raubitsj) [reply wandb/sdk/lib/gitlib.py:133]: we do similar in has_submodule_diff..  so i think without better tests in this codepath id prefer minimal changes.

**Review by raubitsj — APPROVED** [review]
LGTM... but with my reading of this, it feels a bit risky since this codepath is tested very well.

File: wandb/sdk/lib/gitlib.py [hunk 14]
```diff
@@ -23,69 +37,78 @@ def __init__(
...
             return self.repo.git.rev_parse("--show-toplevel")
-        except exc.GitCommandError as e:
+        except GitCommandError as e:
             # todo: collect telemetry on this
             logger.error(f"git root error: {e}")
             return None
 
     @property
-    def dirty(self) -> bool:
+    def dirty(self) -> Any:
```
[the remark this answers is not in the document]
Comment (kptkin) [reply wandb/sdk/lib/gitlib.py:93]: problem with the gitpython typing :(

**Pull request merged**